pkg/kademlia/lookup.go

```diff
@@ -4,89 +4,135 @@
+	client node.Client
+	target dht.NodeID
+	opts   lookupOpts
+
+	cond  *sync.Cond
+	queue *XorQueue
+
+	mu        *sync.Mutex
+	contacted map[string]int
 }
 
-func newSequentialLookup(rt *RoutingTable, nodes []*pb.Node, client node.Client, target dht.NodeID, limit int, bootstrap bool) *sequentialLookup {
-	queue := NewXorQueue(limit)
+func newConcurrentLookup(nodes []*pb.Node, client node.Client, target dht.NodeID, opts lookupOpts) *concurrentLookup {
+	queue := NewXorQueue(opts.concurrency)
 	queue.Insert(target, nodes)
 
-	return &sequentialLookup{
-		contacted:       map[string]bool{},
-		queue:           queue,
-		slowestResponse: 0,
-		client:          client,
-		target:          target,
-		limit:           limit,
-		bootstrap:       bootstrap,
-	}
-}
+	return &concurrentLookup{
+		client: client,
+		target: target,
+		opts:   opts,
 
-func (lookup *sequentialLookup) Run(ctx context.Context) error {
-	for lookup.queue.Len() > 0 {
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		default:
-		}
-
-		next, priority := lookup.queue.Closest()
-		if !lookup.bootstrap && bytes.Equal(priority.Bytes(), make([]byte, len(priority.Bytes()))) {
-			return nil // found the result
-		}
-
-		uncontactedNeighbors := []*pb.Node{}
-		neighbors := lookup.FetchNeighbors(ctx, next)
-		for _, neighbor := range neighbors {
-			if !lookup.contacted[neighbor.GetId()] {
-				uncontactedNeighbors = append(uncontactedNeighbors, neighbor)
-			}
-		}
-		lookup.queue.Insert(lookup.target, uncontactedNeighbors)
+		cond:  &sync.Cond{L: &sync.Mutex{}},
+		queue: queue,
 
-		for lookup.queue.Len() > lookup.limit {
-			lookup.queue.Closest()
-		}
+		mu:        &sync.Mutex{},
+		contacted: map[string]int{},
 	}
-	return nil
 }
 
-func (lookup *sequentialLookup) FetchNeighbors(ctx context.Context, node *pb.Node) []*pb.Node {
-	if node.GetAddress() == nil {
-		return nil
-	}
-	lookup.contacted[node.GetId()] = true
-
-	start := time.Now()
-	neighbors, err := lookup.client.Lookup(ctx, *node, pb.Node{Id: lookup.target.String()})
-	if err != nil {
-		// TODO(coyle): I think we might want to do another look up on this node or update something
-		// but for now let's just log and ignore.
-		log.Printf("Error occurred during lookup for %s on %s :: error = %s", lookup.target.String(), node.GetId(), err.Error())
-		return []*pb.Node{}
-	}
+func (lookup *concurrentLookup) Run(ctx context.Context) error {
+	wg := sync.WaitGroup{}
+
+	// protected by `lookup.cond.L`
+	working := 0
+	allDone := false
+
+	wg.Add(lookup.opts.concurrency)
+	for i := 0; i < lookup.opts.concurrency; i++ {
+		go func() {
+			defer wg.Done()
+			for {
+				var (
+					next     *pb.Node
+					priority big.Int
+				)
+
+				lookup.cond.L.Lock()
+				for {
+					// everything is done, this routine can return
+					if allDone {
+						lookup.cond.L.Unlock()
+						return
+					}
+
+					next, priority = lookup.queue.Closest()
+					if !lookup.opts.bootstrap && priority.Cmp(&big.Int{}) == 0 {
```

Comment:
what is `priority.Cmp(&big.Int{}) == 0` doing ?

Reply:
checking if the priority is 0, which would mean that the xor distance between the target and the node in question is 0. This is a simplification of what was already there: [`bytes.Equal(priority.Bytes(), make([]byte, len(priority.Bytes())))`](https://github.com/storj/storj/blob/master/pkg/kademlia/lookup.go#L51)